modules/dnn/src/onnx/onnx_importer.cpp

```diff
@@ -1957,50 +1957,77 @@ void ONNXImporter::parseGemm(LayerParams& layerParams, const opencv_onnx::NodePr
     addLayer(layerParams, node_proto);
 }
 
-void ONNXImporter::parseMatMul(LayerParams& layerParams, const opencv_onnx::NodeProto& node_proto_)
-{
-    opencv_onnx::NodeProto node_proto = node_proto_;
-    CV_Assert(node_proto.input_size() == 2);
-    layerParams.type = "InnerProduct";
```

Comment:
Why we switch to a new MatMul instead of [GEMM](https://github.com/opencv/opencv/blob/4.x/modules/dnn/src/layers/gemm_layer.cpp)?

Reply:
1. Gemm in dnn handles: 2d x 2d = 2d or 3d (batch) x 2d = 2d (force). MatMul in this PR handles: nd x nd = nd.
2. Broadcasting is handled carefully here in MatMul via `MatMulHelper` so that we don't do physical broadcast.

Reply:
Got it, thanks.